x-pack/plugins/licensing/common/license.ts

```diff
@@ -0,0 +1,155 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { i18n } from '@kbn/i18n';
+import {
+  ElasticsearchError,
+  LicenseType,
+  ILicense,
+  LicenseStatus,
+  LICENSE_CHECK_STATE,
+  LICENSE_TYPE,
+  PublicLicense,
+} from '../server/types';
+
+/**
+ * @public
+ */
+export class License implements ILicense {
+  private readonly license?: PublicLicense['license'];
+  private readonly features?: PublicLicense['features'];
+
+  public error?: ElasticsearchError;
+  public isActive: boolean;
+  public isAvailable: boolean;
+  public isBasic: boolean;
+  public isNotBasic: boolean;
+
+  public uid?: string;
+  public status?: LicenseStatus;
+  public expiryDateInMillis?: number;
+  public type?: LicenseType;
+  public signature: string;
+
+  /**
+   * @internal
+   * Generate a License instance from json representation.
+   */
+  static fromJSON(json: PublicLicense) {
+    return new License(json);
+  }
+
+  constructor({
+    license,
+    features,
+    signature,
+    error,
+  }: PublicLicense & { error?: ElasticsearchError }) {
+    this.isAvailable = Boolean(license);
+    this.license = license;
+    this.features = features;
+    this.error = error;
+    this.signature = signature;
+
+    if (license) {
+      this.uid = license.uid;
+      this.status = license.status;
+      this.expiryDateInMillis = license.expiryDateInMillis;
+      this.type = license.type;
+    }
+
+    this.isActive = this.status === 'active';
+    this.isBasic = this.isActive && this.type === 'basic';
+    this.isNotBasic = this.isActive && this.type !== 'basic';
+  }
+
+  toJSON() {
+    return {
+      license: this.license,
+      features: this.features,
+      signature: this.signature,
+    };
+  }
+
+  public getUnavailableReason() {
+    if (this.error) {
+      if (this.error.status === 400) {
```

Comment:
looks like a leaky abstraction. how about to keep in License error message only?